src/GafferTest/ContextTest.cpp

```diff
@@ -204,3 +204,44 @@ void GafferTest::testEditableScope()
 	}
 
 }
+
+void GafferTest::testNewContextAPIBasics()
+{
+
+	// These are not a very complete tests ( more complete tests are added in Gaffer 0.60
+	// for the actual new code ). But this confirms that the Gaffer 0.59 forward
+	// compatibility shims all compile and call through to the correct functions.
+	ContextPtr baseContext = new Context();
+	GAFFERTEST_ASSERT( baseContext->getIfExists<int>( "a" ) == nullptr );
+	GAFFERTEST_ASSERT( baseContext->getAsData( "a", nullptr ) == nullptr );
+	GAFFERTEST_ASSERT( ((IntData*)baseContext->getAsData( "a", new IntData( 5 ) ).get())->readable() == 5 );
+	try
+	{
+		baseContext->getAsData( "a" );
+		GAFFERTEST_ASSERT( false );
+	}
+	catch( IECore::Exception &e )
+	{
+		// We correctly threw an exception
+	}
+
+	baseContext->set( "a", 10 );
+	GAFFERTEST_ASSERT( *baseContext->getIfExists<int>( "a" ) == 10 );
+	GAFFERTEST_ASSERT( ((IntData*)baseContext->getAsData( "a", nullptr ).get())->readable() == 10 );
+	GAFFERTEST_ASSERT( ((IntData*)baseContext->getAsData( "a", new IntData( 5 ) ).get())->readable() == 10 );
```

Comment:
This leaks `IntData( 5 )` since the argument we take is a raw pointer, and in this case we're not returning it so that it never get assigned to a smart pointer. I think we should be taking `const DataPtr &` here instead? That follows our "pass a smart pointer if the function will modify ownership in some way, otherwise pass a raw pointer" rule.

Reply:
Sure, I guess that's fine.  All known usages of defaultValue in practice pass nullptr anyway, but I guess that's a bit safer.